### PR TITLE
#164892850 social Authentication for facebook  and google

### PR DESCRIPTION
**Pull request by ssanusi**
#### Description
The Pull Request add the feature of authentication on the app via facebook and google 

#### Type of change
 - added Passport-facebook and Passport-google for social authentication

#### How Has This Been Tested?
 - navigate to the route ```/api/v1/auth/facebook``` you will be required to log in to the application 
 - navigate to the route ```/api/v1/auth/google``` you will be required to log in to the application 

#### Checklist:
 - [x] Facebook Authentication
 - [x] Google Authentication
 - [x] Unit Testing
 - [x] Integration Testing

#### PT
[#164892850](https://www.pivotaltracker.com/story/show/164892850)

### CODE REVIEW

File: src/middlewares/socialCallback.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+
+const strategyCallback = async (accessToken, refreshToken, profile, done) => {
+  const {
+    id, displayName, photos, emails
+  } = profile;
+
+  const salt = bcrypt.genSaltSync(10);
+  const hashPassword = bcrypt.hashSync(id, salt);
+  const [user] = await User.findOrCreate({
```
Comment (chikeozulumba) [inline src/middlewares/socialCallback.js:9]: @ssanusi Nice work. Have you considered destructuring to make the code more readable

File: src/server.js [hunk 2]
```diff
@@ -19,6 +21,14 @@ export const startServer = port => new Promise((resolve, reject) => {
     reject(new Error('The server must be started with an available port'));
   }
 
+  passport.serializeUser((user, cb) => {
```
Comment (chikeozulumba) [inline src/server.js:24]: @ssanusi Can shorten the function, as you are just returning the callback function

**Review by chikeozulumba — COMMENTED** [review]
@ssanusi I left some comments on the changed files, kindly review.
Nice work done.

File: src/middlewares/socialCallback.js (hunk 1 above)
Comment (ssanusi) [reply src/middlewares/socialCallback.js:9]: the above user is destructured

File: src/config/passportConfig.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+export default {
+  facebookAuth: {
+    clientID: process.env.FACEBOOK_CLIENT_ID,
+    clientSecret: process.env.FACEBOOK_APP_SECRET,
+    callbackURL: 'http://localhost:4000/api/v1/auth/facebook/callback',
```
Comment (Temmyogunbo) [inline src/config/passportConfig.js]: This won't work in production.

File: src/middleware/socialCallback.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    id, displayName, photos, emails
+  } = profile;
+
+  const salt = bcrypt.genSaltSync(10);
+  const hashPassword = bcrypt.hashSync(id, salt);
+  const [user] = await User.findOrCreate({
+    where: { email: emails[0].value },
+    defaults: {
+      full_name: displayName,
+      password: hashPassword,
```
Comment (Temmyogunbo) [inline src/middleware/socialCallback.js]: Please fix this. Sync with @Raymond-Osy to understand how.

File: src/routers/index.js [hunk 5]
```diff
@@ -26,5 +27,28 @@ router
...
+    res.redirect('/api/v1');
+  },
+);
+
+// Route for google Authentication
+router.get('/auth/google', passport.authenticate('google', { scope: ['email profile'] }));
+router.get(
+  '/auth/google/callback',
+  passport.authenticate('google', { failureRedirect: '/api/v1/auth/login' }),
+  (req, res) => {
```
Comment (Temmyogunbo) [inline src/routers/index.js]: Please research on how best to go about it.

File: src/middlewares/socialCallback.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import bcrypt from 'bcryptjs';
+import db from '../models';
```
Comment (Temmyogunbo) [inline src/middlewares/socialCallback.js]: let us store as models or we destructure the exact model e.g `import {Users} from '../models';
`

**Pull request closed without merging**